Bug: Support for Python 3.12 type keyword

### Description

When upgrading from Python 3.11 syntax to 3.12, my TypeAliases are now `type` lines.

Ex:
```python
LiteralAlias: TypeAlias = Literal['a','b','c']
```
becomes
```python
type LiteralAlias = Literal['a','b','c']
```
With this change Polyfactory no longer recognizes the type:
```
E       polyfactory.exceptions.ParameterException: Unsupported type: LiteralAlias
E       
E       Either extend the providers map or add a factory function for this type.
```

### URL to code causing the issue

_No response_

### MCVE

```python
# Your MCVE code here
```


### Steps to reproduce

```bash
1. Go to '...'
2. Click on '....'
3. Scroll down to '....'
4. See error
```


### Screenshots

"In the format of: `![SCREENSHOT_DESCRIPTION](SCREENSHOT_LINK.png)`"


### Logs

_No response_

### Release Version

2.12.0

### Platform

- [X] Linux
- [ ] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

<!-- POLAR PLEDGE BADGE START -->
---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/polyfactory/issues/455">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/polyfactory/issues/455/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/polyfactory/issues/455/pledge.svg">
</picture>
</a>
<!-- POLAR PLEDGE BADGE END -->

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -34,7 +34,7 @@ jobs:
     strategy:
       fail-fast: true
       matrix:
-        python-version: ["3.8", "3.9", "3.10", "3.11"]
+        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
         pydantic-version: ["1.10", "2.0"]
         sqla-version: ["1.4", "2"]
         exclude:

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -14,13 +14,12 @@ jobs:
 
       - uses: actions/setup-python@v5
         with:
-          python-version: "3.11"
+          python-version: "3.12"
 
       - name: Install Pre-Commit
         run: python -m pip install pre-commit && pre-commit install
 
       - name: Load cached Pre-Commit Dependencies
-        id: cached-pre-commit-dependencies
         uses: actions/cache@v3
         with:
           path: ~/.cache/pre-commit/
@@ -78,15 +77,15 @@ jobs:
         run: echo "PYTHONPATH=$PWD" >> $GITHUB_ENV
 
       - name: Test
-        if: matrix.python-version != '3.11' || matrix.pydantic-version != '2.0' || matrix.sqla-version != '2'
+        if: matrix.python-version != '3.12' || matrix.pydantic-version != '2.0' || matrix.sqla-version != '2'
         run: pdm run pytest tests
 
       - name: Test with Coverage
-        if: matrix.python-version == '3.11' && matrix.pydantic-version == '2.0' && matrix.sqla-version == '2'
+        if: matrix.python-version == '3.12' && matrix.pydantic-version == '2.0' && matrix.sqla-version == '2'
         run: pdm run pytest tests docs/examples --cov=. --cov-report=xml
 
       - uses: actions/upload-artifact@v4
-        if: matrix.python-version == '3.11'
+        if: matrix.python-version == '3.12'
         with:
           name: coverage-xml
           path: coverage.xml
@@ -103,12 +102,12 @@ jobs:
       - name: Set up Python
         uses: actions/setup-python@v5
         with:
-          python-version: "3.11"
+          python-version: "3.12"
 
       - uses: pdm-project/setup-pdm@v3
         name: Set up PDM
         with:
-          python-version: "3.11"
+          python-version: "3.12"
           allow-python-prereleases: true
           cache: true
 

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -20,6 +20,7 @@ jobs:
         run: python -m pip install pre-commit && pre-commit install
 
       - name: Load cached Pre-Commit Dependencies
+        id: cached-pre-commit-dependencies
         uses: actions/cache@v3
         with:
           path: ~/.cache/pre-commit/

--- a/.github/workflows/docs.yaml
+++ b/.github/workflows/docs.yaml
@@ -17,12 +17,12 @@ jobs:
 
       - uses: actions/setup-python@v5
         with:
-          python-version: "3.11"
+          python-version: "3.12"
 
       - uses: pdm-project/setup-pdm@v3
         name: Set up PDM
         with:
-          python-version: "3.11"
+          python-version: "3.12"
           allow-python-prereleases: true
           cache: true
 

--- a/.github/workflows/publish.yaml
+++ b/.github/workflows/publish.yaml
@@ -18,12 +18,12 @@ jobs:
 
       - uses: actions/setup-python@v5
         with:
-          python-version: "3.11"
+          python-version: "3.12"
 
       - uses: pdm-project/setup-pdm@v3
         name: Set up PDM
         with:
-          python-version: "3.11"
+          python-version: "3.12"
           allow-python-prereleases: true
           cache: true
 
@@ -42,12 +42,12 @@ jobs:
 
       - uses: actions/setup-python@v5
         with:
-          python-version: "3.11"
+          python-version: "3.12"
 
       - uses: pdm-project/setup-pdm@v3
         name: Set up PDM
         with:
-          python-version: "3.11"
+          python-version: "3.12"
           allow-python-prereleases: true
           cache: true
 

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -1,5 +1,5 @@
 default_language_version:
-  python: "3.11"
+  python: "3.12"
 repos:
   - repo: https://github.com/compilerla/conventional-pre-commit
     rev: v3.0.0

--- a/polyfactory/utils/helpers.py
+++ b/polyfactory/utils/helpers.py
@@ -8,7 +8,7 @@ try:
 except ImportError:
     NoneType = type(None)  # type: ignore[misc,assignment]
 
-from typing_extensions import get_args, get_origin
+from typing_extensions import TypeAliasType, get_args, get_origin
 
 from polyfactory.constants import TYPE_MAPPING
 from polyfactory.utils.predicates import is_annotated, is_new_type, is_optional, is_safe_subclass, is_union
@@ -65,13 +65,20 @@ def unwrap_annotation(annotation: Any, random: Random) -> Any:
     :returns: The unwrapped annotation.
 
     """
-    while is_optional(annotation) or is_new_type(annotation) or is_annotated(annotation):
+    while (
+        is_optional(annotation)
+        or is_new_type(annotation)
+        or is_annotated(annotation)
+        or isinstance(annotation, TypeAliasType)
+    ):
         if is_new_type(annotation):
             annotation = unwrap_new_type(annotation)
         elif is_optional(annotation):
             annotation = unwrap_optional(annotation)
         elif is_annotated(annotation):
             annotation = unwrap_annotated(annotation, random=random)[0]
+        elif isinstance(annotation, TypeAliasType):
+            annotation = annotation.__value__
 
     return annotation
 

--- a/polyfactory/utils/predicates.py
+++ b/polyfactory/utils/predicates.py
@@ -3,15 +3,7 @@ from __future__ import annotations
 from inspect import isclass
 from typing import Any, Literal, NewType, Optional, TypeVar, Union, get_args
 
-from typing_extensions import (
-    Annotated,
-    NotRequired,
-    ParamSpec,
-    Required,
-    TypeGuard,
-    _AnnotatedAlias,
-    get_origin,
-)
+from typing_extensions import Annotated, NotRequired, ParamSpec, Required, TypeGuard, _AnnotatedAlias, get_origin
 
 from polyfactory.constants import TYPE_MAPPING
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -22,6 +22,7 @@ classifiers = [
     "Programming Language :: Python :: 3.9",
     "Programming Language :: Python :: 3.10",
     "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
     "Programming Language :: Python",
     "Topic :: Software Development :: Libraries",
     "Topic :: Software Development :: Testing :: Unit",

--- a/tests/test_new_types.py
+++ b/tests/test_new_types.py
@@ -1,6 +1,8 @@
+import sys
 from datetime import date
 from decimal import Decimal
-from typing import Any, Dict, List, NewType, Optional, Tuple, Union
+from types import ModuleType
+from typing import Any, Callable, Dict, List, NewType, Optional, Tuple, Union
 
 import pytest
 from pydantic import (
@@ -18,6 +20,7 @@ from pydantic import (
     constr,
 )
 
+from polyfactory.factories.dataclass_factory import DataclassFactory
 from polyfactory.factories.pydantic_factory import ModelFactory
 
 
@@ -137,3 +140,25 @@ def test_constrained_new_types() -> None:
 
     assert isinstance(result.conpositive_float_field, float)
     assert result.conpositive_float_field > 0
+
+
+@pytest.mark.skipif(sys.version_info < (3, 12), reason="3.12 only syntax")
+def test_type_alias(create_module: Callable[[str], ModuleType]) -> None:
+    module = create_module(
+        """
+from typing import Literal
+from dataclasses import dataclass
+
+
+type LiteralAlias = Literal["a", "b"]
+
+
+@dataclass
+class A:
+    field: LiteralAlias
+""",
+    )
+
+    factory = DataclassFactory.create_factory(module.A)  # type: ignore[var-annotated]
+    result = factory.build()
+    assert result.field in {"a", "b"}

